### PR TITLE
fix: Minibar Flag Inconsistency

### LINKED ISSUE #10691: [Bug] Flag display on reviewer minibar is inconsistent when changing flags
###### Reproduction Steps

1. Have `Flag` app bar option to `Always`
2. Open a flagged card
3. Reflag it

https://user-images.githubusercontent.com/69634269/161430479-57b09dd5-dad4-46b5-a4a7-531e488d108a.mp4

###### Expected Result

Flag on minibar should not appear if app bar already has the `Flag` option. Neither the flag should vanish when changing if the former was the intended behavior

###### Actual Result

The minibar flag appear then vanishes when changed

###### Debug info
```
AnkiDroid Version = 2.16alpha54

Android Version = 11

Manufacturer = samsung

Model = SM-N770F

Hardware = exynos9810

Webview User Agent = Mozilla/5.0 (Linux; Android 11; SM-N770F Build/RP1A.200720.012; wv) AppleWebKit/537.36 (KHTML, like Gecko) Version/4.0 Chrome/99.0.4844.88 Mobile Safari/537.36

ACRA UUID = 037bbc86-38a4-4ad7-9c42-05870e9b6086

Scheduler = std2

Crash Reports Enabled = true

DatabaseV2 Enabled = true
```

###### Research
*Enter an [x] character to confirm the points below:*

- [X] I have read the [support page](https://ankidroid.org/docs/help.html) and am reporting a bug or enhancement request specific to AnkiDroid
- [X] I have checked the [manual](https://ankidroid.org/docs/manual.html) and the [FAQ](https://github.com/ankidroid/Anki-Android/wiki/FAQ) and could not find a solution to my issue
- [X] I have searched for similar existing issues here and on the user forum
- [X] (Optional) I have confirmed the issue is not resolved in the latest alpha release ([instructions](https://docs.ankidroid.org/manual.html#betaTesting))

### PR DESCRIPTION
**Pull request by Kshitij09**
Signed-off-by: Kshitij Patil <kshitijpatil98@gmail.com>

## Purpose / Description
Due to some refactorings around `AbstractFlashCardViewer`, especially the revision 66300a61a4c91227423308b6492964a7403dd7eb, which extracts Flag/Mark functionality from this Class, a regression bug was introduced which managed to remain unnoticed for quite a few alpha releases. Since we show certain options on the minibar as per their visibility on AppBar, the order in which we toggle their visibility pre/post invalidating Options Menu, causes this bug. This PR proposes a solution that fixes the ordering issue.

## Fixes
Fixes #10691 

## Approach
The refactor performed in the mentioned commit (66300a61a4c91227423308b6492964a7403dd7eb) moved Flag/Mark toggle responsibility to `Reviewer` class from `AbstractFlashCardViewer`. 
1. While most of the method calls shifted to their respective positions in the subclass, one instance of invoking `drawFlag()` & `drawMark()` in the `onPageFinished()` method of `CardViewerWebClient` wasn't taken care of.2. 
1. A call from `fillFlashCard` to `onFlagChanged()` & `onMarkChanged()` was supporting this expectation to some extent but this wasn't covering all the cases which could invalidate the AppBar Options Menu, rendering the minibar flag/mark inconsistent.
1.  `CardViewerWebClient` actually reacts to an url that starts with `signal:flag_` by calling `executeCommand` method of `AbstractFlashCardViewer` https://github.com/ankidroid/Anki-Android/blob/3404c44d736948bcb7c921ea8e52e598ea7aa11c/AnkiDroid/src/main/java/com/ichi2/anki/AbstractFlashcardViewer.kt#L2280
but since we've removed the commands handling flag actions from this class (moved to `Reviewer`), it wasn't updating the flag state.

I have added a callback that's invoked from `onPageFinished` which provides a hook for `CardViewer` subclasses to perform required actions (Calling `onFlagChanged()` & `onMarkChanged()` in our case).

## How Has This Been Tested?

As of now, by manually running it on a physical device (Pixel 4a)

## Checklist
_Please, go through these checks before submitting the PR._

- [X] You have not changed whitespace unnecessarily (it makes diffs hard to read)
- [X] You have a descriptive commit message with a short title (first line, max 50 chars).
- [X] Your code follows the style of the project (e.g. never omit braces in `if` statements) 
- [X] You have commented your code, particularly in hard-to-understand areas
- [X] You have performed a self-review of your own code
- [ ] UI changes: include screenshots of all affected screens (in particular showing any new or changed strings)
- [ ] UI Changes: You have tested your change using the [Google Accessibility Scanner](https://play.google.com/store/apps/details?id=com.google.android.apps.accessibility.auditor)

### CODE REVIEW

**Comment by Kshitij09** [comment]
@BrayanDSO @david-allison Could you please have a look at this PR once?

File: AnkiDroid/src/main/java/com/ichi2/anki/OnPageFinishedCallback.kt [hunk 1]
```diff
@@ -0,0 +1,8 @@
+package com.ichi2.anki
+
+/**
+ * Intended to be used with [android.webkit.WebViewClient.onPageFinished]
+ */
+interface OnPageFinishedCallback {
```
Comment (david-allison) [inline AnkiDroid/src/main/java/com/ichi2/anki/OnPageFinishedCallback.kt:6]: This /could/ be an open function on `AbstractFlashvardViewer`, but I prefer this architecturally

**Review by david-allison — CHANGES_REQUESTED** [review]
Looks great, pending copyright header

Could you either add a Unit test (which would have previously failed), or a `@NeedsTest` annotation

**Comment by mikehardy** [comment]
@BrayanDSO you have a unique combination in that you noticed quite a few of the menu bar issues and you can build the app quickly from the fork/branch of a PR to check PRs if I remember correctly. It would be really useful to check the effect of this PR on the 7-issue multi-bug that is the app bar state right now...

**Comment by david-allison** [comment]
> Hey, sorry for not getting back on this soon. Where can I find the respective Copyright header for this file? I could see different files have different Years and authors in the header.
> 
> I will think about a unit test for this scenario, but any guidance/reference would be much appreciated

Never any rush. Instructions:

https://github.com/ankidroid/Anki-Android/blob/4745b60d40af443bf0fff4f9d1e39685578ca01b/lint-rules/src/main/java/com/ichi2/anki/lint/rules/CopyrightHeaderExists.java#L61-L72

**Comment by Kshitij09** [comment]
Damm! I was late to notice, I am really shocked how this is working but if you check I have added the callback implementation to `AbstractFlashvardViewer` but haven't really passed `this` reference while creating an instance of `CardViewerWebClient` here https://github.com/ankidroid/Anki-Android/blob/bcae6dc760bc981c7c74ce77490d1445849577b2/AnkiDroid/src/main/java/com/ichi2/anki/AbstractFlashcardViewer.kt#L1021


Would you mind if I fix this by raising another PR with the suggested change?
cc: @david-allison @mikehardy

**Comment by BrayanDSO** [comment]
I'm late to the party. Tested it on main and it looks nice! Thanks Kshitj!

**Pull request merged**